[COMMENTS?] enable easy printing (debugging) of some enum classes

germasch opened the pull request:
I've had this sitting around for a while, and let me start by saying that I don't have strong feelings about it. It adds a somewhat neat feature making it easier to give human-readable output for some enums, but it could also be considered rarely needed bloat.

Essentially, the PR adds `adios::tostring` and overloads for the ostream `operator<<` for enum classes. (It only does it for a couple of them right now, I could add the rest if you want this).  It allows you to do, e.g., 
```cxx
std::cout << variable.ShapeID() << std::endl;
```
and get human readable output.
It gets picked up by googletest, so on an (intentionally failing) test, instead of
```
../testing/adios2/engine/bp/TestBPWriteReadADIOS2.cpp:186: Failure
Expected equality of these values:
  var_i8.ShapeID()
    Which is: 4-byte object <02-00 00-00>
  adios2::ShapeID::LocalArray
    Which is: 4-byte object <05-00 00-00>
```
you get
```
../testing/adios2/engine/bp/TestBPWriteReadADIOS2.cpp:186: Failure
Expected equality of these values:
  var_i8.ShapeID()
    Which is: ShapeID::GlobalArray
  adios2::ShapeID::LocalArray
    Which is: ShapeID::LocalArray
```

In its last commit, it demonstrates that one could also support `std::cout << variable` to do something akin to Python's `__str__` method, which might be useful for debugging, but mostly it's me playing with SFINAE, and I'm less than convinced that this isn't overdoing things.

chuckatkins commented:
Please rebase off current master as the version of clang-format was just bumped from 3.8.1 to 7.0.1

chuckatkins commented on source/adios2/ADIOSTypes.inl:
```diff
@@ -111,6 +111,12 @@ struct TypeInfo<
+{
+    return os << ToString(value);
+}
+
```
Forgive me if I'm not seeing it but what the the `Enable` buy us here that you wouldn't get from the implementation resolving the `ToString` signature?

germasch replied:
I will look at that again, but I think the `enable_if` may really be required there in order to not overload existing standard library `operator<<` overloads.

chuckatkins commented:
So, in short, I really like the capability and think it's super useful.  Also the implementation via `ToString` with `operator<<` is good.  I think there's a few issues with the implementation though that are over-complicating things and can be simplified.

germasch commented on source/adios2/ADIOSTypes.cpp:
```diff
@@ -0,0 +1,48 @@
+std::string ToString(IOMode value) { return MapIOMode2String.at(value); }
+
+std::string ToString(Mode value) { return MapMode2String.at(value); }
+
```
Sure. I'll take the opportunity to handle the case of an unknown value being passed by returning a corresponding message rather than throwing an exception.

chuckatkins replied:
Something as simple as:
```c++
switch(value)
{
    case IOMode::Independent : return "IOMode::Independent";
    case IOMode::Collective: return "IOMode::Collective";
    default: throw std::runtime_error("Invalid IOMode");
}
```
Should work and then you remove the need for the map.

germasch commented on bindings/CXX11/cxx11/Variable.h:
```diff
@@ -304,6 +304,23 @@ class Variable
+
+template <typename T,
+          typename std::enable_if<IsVariable<T>::value, int>::type = 0>
+std::string ToString(T value);
```
Yes, I think that should do. Not sure why I did the `enable_if` there, I think you're right that it should work without.

germasch commented:
Thanks, I appreciate that. Thanks for the detailed review, too. I'll address the issues above and add some missing pieces to make this PR more complete.

williamfgc commented:
@germasch @chuckatkins thanks, can we also add some documentation in the user guide under  https://github.com/ornladios/ADIOS2/blob/master/docs/user_guide/source/components/variable.rst Since this is a new feature, it would be helpful to add this info for those wanting to try. Thanks!

germasch commented:
I added more commits:
* Avoid the over-complicated `enable_if` for `ToString(const Variable<T>&)` as @chuckatkins pointed out
* Use switch/case instead of `std::map`. In the `default` branch, I just return a string indicating that the enum value is not known now. I did this before I saw @chuckatkins comment about throwing an exception there. I'll be happy to change it to throwing, if you prefer. My thinking is that this functionality will probably be mostly used for debugging, and in case someone adds a new enum value but forgets to add the string translation for it, it's potentially confusing to have a new exception be thrown as a consequence of your debugging printf (if I may call it that, I guess I should say "debugging std::cerr"). 
* Add support for printing the remaining enums, and the remaining cxx11 binding objects.
* Add tests.
* Add some docs (I opted to only mention `ToString` and the `ostream& operator<<` in one place, and added an example about its use.
* I also added support for `adios2::Dims`, but this ends up being a bit half-baked, because `adios2::Dims` is just a type alias, and the underlying `std::vector<size_t>` doesn't live in the `adios2::` namespace. Hence ADL doesn't work, so one has to fully qualify `adios2::ToString` to use it, and the ostream version just doesn't work at all.

germasch commented:
In the mean time, ignore my force-pushing while trying to figure out why MSVC is unhappy...

germasch commented:
Alright, so of course even seemingly easy stuff always hits some hiccups.

First, just to be explicit about it, this PR changes the CXX11 API by adding `IO::Name()` which does the obvious thing, and which is needed for its `ToString` overload. It's just an addition, there should be issue with it.

Second, it now changes the `ADIOS` constructors to be explicit. This is to work around what I think is a compiler bug in MSVC15/17, but I think it's also generally a good idea, following "C.46: By default, declare single-argument constructors explicit" from the [C++ core guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-explicit). By means of the default arguments, `ADIOS` essentially has single argument constructors, and this change prohibits, e.g.,
```cxx
ADIOS ad = "config.xml";
ADIOS ad = MPI_COMM_WORLD;
```
which doesn't seem like a good idea in the first place, though it is supported right now.

germasch commented:
That gives me a chance to provide an answer to @chuckatkins's question with regard to
```cxx
template <typename T, typename Enable = decltype(ToString(std::declval<T>()))>
std::ostream &operator<<(std::ostream &os, const T &value);
```
in more detail. The reason that the Enable SFINAE thing is required there is that otherwise one ends up
with ambiguous overloads of `operator<<(ostream&, ...)`. (Maybe it's possible to do the SFINAE thing in a different, clearer fashion -- I'd be happy to take suggestions.)

Without the `Enable`, this would create an overload of `operator<<` in the `adios2` namespace for every possible type `T`. If one does something simple like `std::cout << "abc"` while the `adios2` namespace is in scope, one ends up with two conflicting overloads, one from the standard library, which is intended, and the one from `adios2`. This typically makes thing fail because the compiler recognizes the ambiguity, but even in a situation where only the adios2 overload is possible, it will then fail to compile if no `ToString` overload for the particular type exists.

So the solution to this, and I believe that's the standard handling of this type of situation, is to take `adios2::operator<<` out of the overload resolution using SFINAE. That's what the `Enable` part above does, it pretends to create an instance of `T`, passes it to `ToString` and determines the return type. If `ToString` doesn't take a `T`, that expression will fail and the overload will not be considered.

So that's all good in theory, but in practice MSVC had some problem with it. I don't have local access to MSVC 15/17, and I tried to reproduce the problem on MSVC19 on godbolt.org, but could not reproduce it there (maybe it's fixed in that version of the compiler). Still, I believe the issue is a compiler bug: If you do `std::cout << "abc"`, MSVC considers the `adios2::operator<<` overload valid for `char[3]` ("abc"). My guess is that it does so via converting `char [3]` -> `std::string` -> `ADIOS`, the last part because of the `ADIOS(const std::string& config = "", bool Debug = DebugON)` constructor. It does look like a compiler bug to me, because it considers two implicit conversions, while only one should be allowed.

But anyway, making the constructors explicit prevents the conversion chain above, so the tests pass on MSVC, and I think that's generally a good change.

germasch commented:
So as if this wasn't complicated enough yet, I have one more consideration: I named the function `adios2::ToString` to be inline with adios2 naming conventions, but essentially thinking of `std::to_string`.  There is a question whether it would better be named `adios2::to_string` to be consistent with the standard library. It would allow to do things similarly to how `swap` is supposed to be used, so that [ADL](https://en.cppreference.com/w/cpp/language/adl) can do its magic:

```cxx
template <class T>
std::string whatever(const T& t)
{
    using std::to_string;
    // to_string will now work on, e.g., T == double, as well as T == adios2::Engine
    throw std::runtime_error("value " + to_string(value) + " not supported");
}
```

Not really a likely use case, but since `ToString` is exposed as part of the user-visible API, now is the time to get the naming right...

germasch commented:
superseded by #1385
